tests/core/query/test_relational_query.py

```diff
@@ -23,6 +23,110 @@ def test_match_table_to_element(sdata_query_aggregation):
+    assert all(element_dict[key] is None for key in element_dict)
+    assert all(table.obs.index == ["7", "8", "19", "20"])
+
+    with pytest.warns(UserWarning, match="The element"):
+        element_dict, table = join_sdata_spatialelement_table(
+            sdata_query_aggregation, ["values_circles", "values_polygons", "by_polygons"], "table", "inner"
+        )
+    indices = pd.concat(
+        [element_dict["values_circles"].index.to_series(), element_dict["values_polygons"].index.to_series()]
+    )
+    assert all(table.obs["instance_id"] == indices.values)
+    assert element_dict["by_polygons"] is None
+
+
+def test_join_spatialelement_table_fail(full_sdata):
+    with pytest.warns(UserWarning, match="Images:"):
+        join_sdata_spatialelement_table(full_sdata, ["image2d", "labels2d"], "table", "left_exclusive")
+    with pytest.warns(UserWarning, match="Tables:"):
+        join_sdata_spatialelement_table(full_sdata, ["labels2d", "table"], "table", "left_exclusive")
+    with pytest.raises(ValueError, match="`not_join` is not a"):
+        join_sdata_spatialelement_table(full_sdata, "labels2d", "table", "not_join")
+
+
+def test_left_exclusive_and_right_join(sdata_query_aggregation):
+    # Test case in which all table rows match rows in elements
+    element_dict, table = join_sdata_spatialelement_table(
+        sdata_query_aggregation, ["values_circles", "values_polygons"], "table", "left_exclusive"
+    )
+    assert all(element_dict[key] is None for key in element_dict)
+    assert table is None
+
+    # Dropped indices correspond to instance ids 7, 8, 10 and 11
+    sdata_query_aggregation["table"] = sdata_query_aggregation["table"][
+        sdata_query_aggregation["table"].obs.index.drop(["7", "8", "19", "20"])
+    ]
+    with pytest.warns(UserWarning, match="The element"):
+        element_dict, table = join_sdata_spatialelement_table(
+            sdata_query_aggregation, ["values_polygons", "by_polygons"], "table", "left_exclusive"
+        )
+    assert table is None
+    assert not set(element_dict["values_polygons"].index).issubset(sdata_query_aggregation["table"].obs["instance_id"])
+
+    # test right join
+    with pytest.warns(UserWarning, match="The element"):
+        element_dict, table = join_sdata_spatialelement_table(
+            sdata_query_aggregation, ["values_circles", "values_polygons", "by_polygons"], "table", "right"
+        )
+    assert table is sdata_query_aggregation["table"]
+    assert not {7, 8}.issubset(element_dict["values_circles"].index)
+    assert not {10, 11}.issubset(element_dict["values_polygons"].index)
+
+    element_dict, table = join_sdata_spatialelement_table(
+        sdata_query_aggregation, ["values_circles", "values_polygons"], "table", "left_exclusive"
+    )
+    assert table is None
+    assert not np.array_equal(
+        sdata_query_aggregation["table"].obs.iloc[7:9, 1].values, element_dict["values_circles"].index.values
+    )
+    assert not np.array_equal(
```

Comment:
I also would replace this test. Here the LHS is by construction an empty series, so it will be almost always different from the RHS. I would instead verify that the RHS has the indices that we removed `array([10, 11])`.

Reply:
but because the table is annotating multiple elements index of `10` and `11` is not actually `10` and `11` in the original table. Instead it is `19` and `20`. It also tests this way that the original table in the `SpatialData` object stays the same.

Reply:
just for the record we are not overwriting elements in the `Spatialdata` object when doing joins.

Reply:
Same as for above, you are correct.